File: libptiff/src/compression/lzw.cpp
#include <cstdint>
#include <map>
#include <utility>

#include <ptiff/compression/lzw.hpp>

namespace ptiff::compression {

namespace {

constexpr std::uint16_t clearCode = 256;
constexpr std::uint16_t eoiCode = 257;
constexpr std::uint16_t firstDictionaryCode = 258;
constexpr std::uint16_t maxDictionaryCode = 4094;

/// TIFF's "early change" code-width schedule: width grows one code sooner than a naive scheme
/// (at 511/1023/2047 instead of 512/1024/2048).
constexpr std::uint8_t codeWidthFor(std::uint16_t nextCode) noexcept {
    if (nextCode >= 2047) {
        return 12;
    }
    if (nextCode >= 1023) {
        return 11;
    }
    if (nextCode >= 511) {
        return 10;
    }
    return 9;
}

using Table = std::map<std::pair<std::vector<std::byte>, std::byte>, std::uint16_t>;

// Returns the LZW code for a phrase: its own byte for a 1-byte phrase (codes 0..255, which the
// decoder seeds implicitly), or the table code assigned by the encoder for longer phrases
// (258..4094). Seq is always a phrase the encoder previously inserted in its `Table`, so the
// table's `->second` gives the code. The single-byte case maps to the byte value directly
// (codes 0..255); nothing else in this encoder emits a code below 258.
std::uint16_t codeForByteSequence(const std::vector<std::byte>& seq, const Table& table) {
    if (seq.size() == 1) {
        return static_cast<std::uint8_t>(std::to_integer<std::uint8_t>(seq[0]));
    }
    // Longer phrases: rebuild the (prefix, lastByte) key that insert stored, and fetch its code.
    const std::vector<std::byte> prefix{seq.begin(), seq.end() - 1};
    return table.at(std::make_pair(prefix, seq.back()));
}

// Packs `codes` MSB-first, honoring the decoder's width transitions. The decoder reads each
// code with its current `codeWidth`, and only grows the width from the 4th dictionary entry
// AFTER a non-clear/non-EOI code (via codeWidthFor(nextCode)), AND only when it already holds an
// old entry (`haveOldEntry`) — i.e. never for the first normal code after a clear. So the width
// used for output code i is computed by replaying the same state machine: start after a clear at
// 9 bits; a normal code (not clear/EOI) at table-size S uses `codeWidthFor(S)`, and only if we
// already emitted a previous normal code since the last clear does it then advance the table size
// by 1. This mirrors the decoder's `haveOldEntry` gating so the width schedule stays in sync at
// the early-change boundaries (511/1023/2047). This function re-derives that width per index in
// one forward pass over the code list.
Result<std::vector<std::byte>> writeCodesToBits(const std::vector<std::uint32_t>& codes) {
    std::vector<std::byte> out;
    std::uint32_t bitsAcc = 0; // pending bits, MSB-aligned in the accumulator
    unsigned bitsPending = 0;
    std::uint16_t tableSize = firstDictionaryCode; // 258 after a fresh clear
    bool havePrevious = false;                     // mirrors the decoder's haveOldEntry gating
    auto width = [&]() { return codeWidthFor(tableSize); };

    auto flushByte = [&]() {
        out.push_back(static_cast<std::byte>((bitsAcc >> (bitsPending - 8)) & 0xFFU));
        bitsPending -= 8;
    };

    for (const auto code : codes) {
        const std::uint8_t w = width(); // decoder's width when reading this code
        bitsAcc = (bitsAcc << w) | (code & ((1U << w) - 1U));
        bitsPending += w;
        while (bitsPending >= 8)
            flushByte();

        if (code == clearCode) {
            tableSize = firstDictionaryCode; // decoder resets to 9 bits on a clear
            havePrevious = false;            // ...and forgets its old entry too
        } else if (code != eoiCode) {
            // Only grow the table once we already hold a previous normal code, exactly as the
            // decoder grows nextCode only when haveOldEntry is true (never for the first normal
            // code after a clear).
            if (havePrevious && tableSize < maxDictionaryCode) {
                ++tableSize;
            }
            havePrevious = true;
        }
    }
    // Emit any trailing zero-padded partial byte so the stream is byte-aligned (the decoder
    // stops at EOI, ignoring trailing padding).
    if (bitsPending > 0) {
        out.push_back(static_cast<std::byte>((bitsAcc << (8 - bitsPending)) & 0xFFU));
    }
    return out;
}

} // namespace

Result<std::vector<std::byte>> decodeLzw(std::span<const std::byte> input,
                                         std::size_t expectedSize) {
    std::vector<std::vector<std::byte>> table(firstDictionaryCode);
    for (int i = 0; i < 256; ++i) {
        table[static_cast<std::size_t>(i)] = {static_cast<std::byte>(i)};
    }

    std::size_t bitPos = 0;
    const std::size_t totalBits = input.size() * 8;

    auto readCode = [&](std::uint8_t width) -> Result<std::uint16_t> {
        if (bitPos + width > totalBits) {
            return std::unexpected(
                Error{ErrorCode::InvalidArgument, "decodeLzw: bitstream exhausted before EOI"});
        }
        std::uint16_t code = 0;
        for (std::uint8_t i = 0; i < width; ++i) {
            const std::size_t byteIndex = bitPos / 8;
            const std::size_t bitIndex = 7 - (bitPos % 8);
            const auto bit = (std::to_integer<std::uint8_t>(input[byteIndex]) >> bitIndex) & 1U;
            code = static_cast<std::uint16_t>((code << 1) | static_cast<std::uint16_t>(bit));
            ++bitPos;
        }
        return code;
    };

    std::vector<std::byte> output;
    output.reserve(expectedSize);
    std::uint16_t nextCode = firstDictionaryCode;
    std::uint8_t codeWidth = 9;
    bool haveOldEntry = false;
    std::vector<std::byte> oldEntry;

    auto appendBounded = [&](const std::vector<std::byte>& entry) -> Result<void> {
        if (output.size() + entry.size() > expectedSize) {
            return std::unexpected(Error{ErrorCode::InvalidArgument,
                                         "decodeLzw: decoded output exceeds expectedSize"});
        }
        output.insert(output.end(), entry.begin(), entry.end());
        return {};
    };

    while (true) {
        auto code = readCode(codeWidth);
        if (!code.has_value()) {
            return std::unexpected(code.error());
        }

        if (*code == clearCode) {
            table.resize(firstDictionaryCode);
            nextCode = firstDictionaryCode;
            codeWidth = 9;
            haveOldEntry = false;
            continue;
        }
        if (*code == eoiCode) {
            break;
        }

        std::vector<std::byte> entry;
        if (*code < 256) {
            entry = table[*code];
        } else if (*code >= firstDictionaryCode && *code < nextCode) {
            entry = table[*code];
        } else if (*code == nextCode && haveOldEntry) {
            entry = oldEntry;
            entry.push_back(oldEntry.front());
        } else {
            return std::unexpected(
                Error{ErrorCode::InvalidArgument, "decodeLzw: invalid code in compressed stream"});
        }

        auto appended = appendBounded(entry);
        if (!appended.has_value()) {
            return std::unexpected(appended.error());
        }

        if (haveOldEntry) {
            if (nextCode > maxDictionaryCode) {
                return std::unexpected(Error{ErrorCode::InvalidArgument,
                                             "decodeLzw: dictionary exceeded 12-bit table"});
            }
            std::vector<std::byte> newEntry = oldEntry;
            newEntry.push_back(entry.front());
            if (static_cast<std::size_t>(nextCode) >= table.size()) {
                table.resize(static_cast<std::size_t>(nextCode) + 1);
            }
            table[nextCode] = std::move(newEntry);
            ++nextCode;
            codeWidth = codeWidthFor(nextCode);
        }

        oldEntry = entry;
        haveOldEntry = true;
    }

    if (output.size() != expectedSize) {
        return std::unexpected(Error{ErrorCode::InvalidArgument,
                                     "decodeLzw: decoded output shorter than expectedSize"});
    }
    return output;
}

Result<std::vector<std::byte>> encodeLzw(std::span<const std::byte> input) {
    if (input.size() > 0xFFFFFFFFU) {
        return std::unexpected(
            Error{ErrorCode::InvalidArgument, "encodeLzw: input exceeds uint32 strip byte count"});
    }

    std::vector<std::uint32_t> codes;
    codes.push_back(clearCode); // start with a clear so decoders reset their table

    // The first 256 single-byte entries are implicit in the decoder (table[0..255] = the byte).
    // Store phrases (dictionaryStart, lookupByte) -> code. A naive map over growing byte vectors is
    // correct but O(n^2); an acceptable simplification here given the small strips this backend
    // writes. The existing decoder only requires the emitted codes to be valid; correctness over
    // speed is the goal.
    Table table;

    std::size_t pos = 0;
    std::uint16_t nextCode = firstDictionaryCode; // 258
    std::uint8_t codeWidth = 9;
    auto resetToFull = [&]() {
        nextCode = firstDictionaryCode;
        codeWidth = 9;
        table.clear();
    };

    std::vector<std::byte> current;
    if (pos < input.size())
        current.push_back(input[pos++]);

    while (pos < input.size()) {
        const std::byte next = input[pos];
        std::pair<std::vector<std::byte>, std::byte> key{current, next};
        const auto it = table.find(key);
        if (it != table.end()) {
            current.push_back(next); // extend the current phrase
            ++pos;
        } else {
            // Emit the code for `current` at the current width.
            codes.push_back(codeForByteSequence(current, table));
            // Add the newly discovered phrase (current + next) to the table.
            if (nextCode < maxDictionaryCode) {
                table.emplace(std::move(key), nextCode);
                ++nextCode;
                if (nextCode > 2046)
                    codeWidth = 12;
                else if (nextCode > 1022)
                    codeWidth = 11;
                else if (nextCode > 510)
                    codeWidth = 10;
                if (nextCode >= maxDictionaryCode) {
                    // table is full: emit clear, reset, then start a fresh current phrase
                    codes.push_back(clearCode);
                    resetToFull();
                }
            }
            current.assign(1, next);
            ++pos;
        }
    }
    if (!current.empty()) {
        codes.push_back(codeForByteSequence(current, table));
    }
    codes.push_back(eoiCode);

    // Convert the code list to an MSB-first bitstream, growing width per the early-change table,
    // computed from the position in the code list the same way decodeLzw derives it (width used for
    // a code is codeWidthFor(nextCodeBeforeAddingThatCode)); zero-pad the final partial byte.
    return writeCodesToBits(codes);
}

} // namespace ptiff::compression

File: crates/ptiff-core/src/io/backend/tiff/compression/lzw.rs
//! TIFF-variant LZW compression (Compression = 5).
//!
//! Mirrors `ptiff::compression::lzw.{hpp,cpp}`.

use std::collections::BTreeMap;

use crate::{Error, Result};

const CLEAR_CODE: u16 = 256;
const EOI_CODE: u16 = 257;
const FIRST_DICTIONARY_CODE: u16 = 258;
const MAX_DICTIONARY_CODE: u16 = 4094;

/// TIFF's "early change" code-width schedule: width grows one code sooner than
/// a naive scheme (at 511/1023/2047 instead of 512/1024/2048).
#[must_use]
const fn code_width_for(next_code: u16) -> u8 {
    if next_code >= 2047 {
        12
    } else if next_code >= 1023 {
        11
    } else if next_code >= 511 {
        10
    } else {
        9
    }
}

type Table = BTreeMap<(Vec<u8>, u8), u16>;

/// Returns the LZW code for a phrase: its own byte for a 1-byte phrase (codes
/// 0..255), or the table code assigned by the encoder for longer phrases
/// (258..4094).
fn code_for_byte_sequence(seq: &[u8], table: &Table) -> u16 {
    if seq.len() == 1 {
        return u16::from(seq[0]);
    }
    let prefix = seq[..seq.len() - 1].to_vec();
    table[&(prefix, seq[seq.len() - 1])]
}

/// Packs `codes` MSB-first, honoring the decoder's width transitions.
fn write_codes_to_bits(codes: &[u32]) -> Vec<u8> {
    let mut out = Vec::new();
    let mut bits_acc: u32 = 0;
    let mut bits_pending: u32 = 0;
    let mut table_size: u16 = FIRST_DICTIONARY_CODE;
    let mut have_previous = false;

    let read_width = |table_size: u16| code_width_for(table_size);

    for &code in codes {
        let w = read_width(table_size) as u32;
        let mask = (1u32 << w) - 1;
        bits_acc = (bits_acc << w) | (code & mask);
        bits_pending += w;
        while bits_pending >= 8 {
            let byte = ((bits_acc >> (bits_pending - 8)) & 0xFF) as u8;
            out.push(byte);
            bits_pending -= 8;
        }

        if code == u32::from(CLEAR_CODE) {
            table_size = FIRST_DICTIONARY_CODE;
            have_previous = false;
        } else if code != u32::from(EOI_CODE) {
            // Exactly mirrors the C++ oracle's bit-writer: the table grows only
            // once we already hold a previous normal code (mirrors the decoder's
            // `have_old_entry` gating), but `have_previous` is set unconditionally
            // for any non-clear/non-EOI code. Without the unconditional set the
            // table never grows, producing a stream the standard LZW decoder
            // (e.g. weezl/image-tiff) rejects.
            if have_previous && table_size < MAX_DICTIONARY_CODE {
                table_size += 1;
            }
            have_previous = true;
        }
    }
    // Emit any trailing zero-padded partial byte.
    if bits_pending > 0 {
        out.push(((bits_acc << (8 - bits_pending)) & 0xFF) as u8);
    }
    out
}

/// Decodes TIFF-variant LZW-compressed data: MSB-first bit packing, 9-bit codes
/// growing to 12 bits, clear code 256 and end-of-information code 257, and
/// TIFF's "early change" quirk (the code width grows one code earlier than
/// standard LZW/GIF). Decoding stops the instant exactly `expected_size` bytes
/// have been produced.
///
/// # Errors
///
/// Returns [`crate::ErrorCode::InvalidArgument`] if a code references a
/// dictionary entry that doesn't exist, decoding would produce more than
/// `expected_size` bytes (decompression-bomb guard), the bitstream is exhausted
/// before an EOI code appears, or the decoded output is shorter than
/// `expected_size` once EOI is reached.
pub fn decode_lzw(input: &[u8], expected_size: usize) -> Result<Vec<u8>> {
    let mut table: Vec<Vec<u8>> = (0..256u32).map(|i| vec![i as u8]).collect();

    let mut bit_pos: usize = 0;
    let total_bits = input.len() * 8;

    let mut read_code = |width: u8, input: &[u8]| -> Result<u16> {
        let w = usize::from(width);
        if bit_pos + w > total_bits {
            return Err(Error::invalid_argument(
                "decodeLzw: bitstream exhausted before EOI",
            ));
        }
        let mut code: u16 = 0;
        for _ in 0..w {
            let byte_index = bit_pos / 8;
            let bit_index = 7 - (bit_pos % 8);
            let bit = (u16::from(input[byte_index]) >> bit_index) & 1;
            code = (code << 1) | bit;
            bit_pos += 1;
        }
        Ok(code)
    };

    let mut output = Vec::with_capacity(expected_size);
    let mut next_code: u16 = FIRST_DICTIONARY_CODE;
    let mut code_width: u8 = 9;
    let mut have_old_entry = false;
    let mut old_entry: Vec<u8> = Vec::new();

    loop {
        let code = read_code(code_width, input)?;

        if code == CLEAR_CODE {
            table.resize(FIRST_DICTIONARY_CODE as usize, Vec::new());
            next_code = FIRST_DICTIONARY_CODE;
            code_width = 9;
            have_old_entry = false;
            continue;
        }
        if code == EOI_CODE {
            break;
        }

        let entry: Vec<u8> =
            if u32::from(code) < 256 || (code >= FIRST_DICTIONARY_CODE && code < next_code) {
                table[code as usize].clone()
            } else if code == next_code && have_old_entry {
                let mut e = old_entry.clone();
                e.push(old_entry[0]);
                e
            } else {
                return Err(Error::invalid_argument(
                    "decodeLzw: invalid code in compressed stream",
                ));
            };

        if output.len() + entry.len() > expected_size {
            return Err(Error::invalid_argument(
                "decodeLzw: decoded output exceeds expectedSize",
            ));
        }
        output.extend_from_slice(&entry);

        if have_old_entry {
            if next_code > MAX_DICTIONARY_CODE {
                return Err(Error::invalid_argument(
                    "decodeLzw: dictionary exceeded 12-bit table",
                ));
            }
            let mut new_entry = old_entry.clone();
            new_entry.push(*entry.first().unwrap_or(&0));
            if next_code as usize >= table.len() {
                table.resize(next_code as usize + 1, Vec::new());
            }
            table[next_code as usize] = new_entry;
            next_code += 1;
            code_width = code_width_for(next_code);
        }

        old_entry = entry;
        have_old_entry = true;
    }

    if output.len() != expected_size {
        return Err(Error::invalid_argument(
            "decodeLzw: decoded output shorter than expectedSize",
        ));
    }
    Ok(output)
}

/// Encodes `input` with the TIFF-variant LZW, inverse of [`decode_lzw`]. Emits
/// an initial clear code 256, then greedy dictionary matches; the dictionary
/// grows from code 258 toward 4094 and is cleared when full. Codes are MSB-first
/// with the TIFF "early change" width schedule.
///
/// # Errors
///
/// Returns [`crate::ErrorCode::InvalidArgument`] if `input.len()` exceeds
/// `u32::MAX` (the strip's `u32` byte count).
pub fn encode_lzw(input: &[u8]) -> Result<Vec<u8>> {
    if input.len() as u64 > u32::MAX as u64 {
        return Err(Error::invalid_argument(
            "encodeLzw: input exceeds uint32 strip byte count",
        ));
    }

    let mut codes: Vec<u32> = vec![u32::from(CLEAR_CODE)];
    let mut table: Table = BTreeMap::new();
    let mut next_code: u16 = FIRST_DICTIONARY_CODE;

    let reset_to_full = |table: &mut Table, next_code: &mut u16| {
        *next_code = FIRST_DICTIONARY_CODE;
        table.clear();
    };

    let n = input.len();
    let mut pos = 0usize;
    let mut current: Vec<u8> = Vec::new();
    if pos < n {
        current.push(input[pos]);
        pos += 1;
    }

    while pos < n {
        let next = input[pos];
        let key = (current.clone(), next);
        if table.contains_key(&key) {
            current.push(next);
            pos += 1;
        } else {
            codes.push(u32::from(code_for_byte_sequence(&current, &table)));
            if next_code < MAX_DICTIONARY_CODE {
                table.insert(key, next_code);
                next_code += 1;
                if next_code >= MAX_DICTIONARY_CODE {
                    codes.push(u32::from(CLEAR_CODE));
                    reset_to_full(&mut table, &mut next_code);
                }
            }
            current = vec![next];
            pos += 1;
        }
    }
    if !current.is_empty() {
        codes.push(u32::from(code_for_byte_sequence(&current, &table)));
    }
    codes.push(u32::from(EOI_CODE));

    Ok(write_codes_to_bits(&codes))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn encode_then_decode_round_trips() {
        // A small but not-trivial strip with repeated patterns.
        let data: Vec<u8> = b"abcabcabcabcXYZXYZXYZ".to_vec();
        let encoded = encode_lzw(&data).unwrap();
        let decoded = decode_lzw(&encoded, data.len()).unwrap();
        assert_eq!(decoded, data);
    }

    #[test]
    fn encode_then_decode_round_trips_binary() {
        // Repetitive binary data exercises the dictionary well.
        let mut data = vec![0u8; 128];
        for i in 0..64 {
            data[i] = 0xAB;
            data[64 + i] = (i * 7) as u8;
        }
        let encoded = encode_lzw(&data).unwrap();
        let decoded = decode_lzw(&encoded, data.len()).unwrap();
        assert_eq!(decoded, data);
    }

    #[test]
    fn encode_single_byte() {
        let data = vec![0x42u8];
        let encoded = encode_lzw(&data).unwrap();
        let decoded = decode_lzw(&encoded, data.len()).unwrap();
        assert_eq!(decoded, data);
    }

    #[test]
    fn decode_rejects_expected_size_bomb() {
        // A short stream cannot produce more than its contents.
        let data: Vec<u8> = b"hello".to_vec();
        let encoded = encode_lzw(&data).unwrap();
        // Asking for many more bytes than decoded must fail the bomb guard.
        assert!(decode_lzw(&encoded, 10_000).is_err());
    }
}
